db/seqno_to_time_mapping.cc

```diff
@@ -0,0 +1,298 @@
+      return s;
+    }
+    val.Add(base);
+    seqno_time_mapping_.emplace_back(val);
+    base = val;
+  }
+  return Status::OK();
+}
+
+void SeqnoToTimeMapping::SeqnoToTimePair::Encode(std::string& dest) const {
+  PutVarint64Varint64(&dest, seqno, time);
+}
+
+Status SeqnoToTimeMapping::SeqnoToTimePair::Decode(Slice& input) {
+  if (!GetVarint64(&input, &seqno)) {
+    return Status::Corruption("Invalid sequence number");
+  }
+  if (!GetVarint64(&input, &time)) {
+    return Status::Corruption("Invalid time");
+  }
+  return Status::OK();
+}
+
+bool SeqnoToTimeMapping::Append(SequenceNumber seqno, uint64_t time,
+                                SequenceNumber earliest_mem_seqno) {
+  assert(is_sorted_);
+
+  // take the hint from caller and handle earliest_mem_seqno first
+  if (earliest_mem_seqno > 0) {
+    auto it = std::upper_bound(seqno_time_mapping_.begin(),
+                               seqno_time_mapping_.end(), earliest_mem_seqno);
+    if (it != seqno_time_mapping_.begin()) {
+      it--;
+      seqno_time_mapping_.erase(seqno_time_mapping_.begin(), it);
+    }
+  }
+
+  // skip seq number 0, which may have special meaning, like zeroed out data
+  if (seqno == 0) {
+    return false;
+  }
+  if (!Empty()) {
+    if (seqno < Last().seqno || time < Last().time) {
+      return false;
+    }
+    if (seqno == Last().seqno) {
+      Last().time = time;
+      return true;
+    }
+    if (time == Last().time) {
+      // new sequence has the same time as old one, no need to add new mapping
+      return false;
+    }
+  }
+
+  seqno_time_mapping_.emplace_back(seqno, time);
+
+  if (seqno_time_mapping_.size() > max_capacity_) {
+    seqno_time_mapping_.pop_front();
+  }
+  return true;
+}
+
+bool SeqnoToTimeMapping::Resize(uint64_t min_time_duration,
+                                uint64_t max_time_duration) {
+  uint64_t new_max_capacity =
+      CalculateMaxCapacity(min_time_duration, max_time_duration);
+  if (new_max_capacity == max_capacity_) {
+    return false;
+  } else if (new_max_capacity < seqno_time_mapping_.size()) {
+    uint64_t delta = seqno_time_mapping_.size() - new_max_capacity;
+    seqno_time_mapping_.erase(seqno_time_mapping_.begin(),
+                              seqno_time_mapping_.begin() + delta);
+  }
+  max_capacity_ = new_max_capacity;
+  return true;
+}
+
+Status SeqnoToTimeMapping::Sort() {
+  if (is_sorted_ || seqno_time_mapping_.empty()) {
+    return Status::OK();
+  }
+
+  std::deque<SeqnoToTimePair> copy = std::move(seqno_time_mapping_);
+
+  std::sort(copy.begin(), copy.end());
+
+  seqno_time_mapping_.clear();
+
+  // remove seqno = 0, which may have special meaning, like zeroed out data
+  while (copy.front().seqno == 0) {
+    copy.pop_front();
+  }
+
+  SeqnoToTimePair prev = copy.front();
+  for (const auto& it : copy) {
+    // If sequence number is the same, pick the one with larger time, which is
+    // more accurate than the older time.
+    if (it.seqno == prev.seqno) {
```

Comment:
Since duplication should be very common (since compaction input files from non-L0 tend to have the same source of seqno->time mapping), I wonder whether we should do Sort(), or always maintain deduplicated order in Append().

Reply:
`Append()` does maintain the order and dedup. (which is used for flush).
`Sort()` is mostly needed in compaction, which add all information together and do sort once. (ideally we should do k-merge sort, but just to keep it simple and also make it easier to add smallest_seqno -> oldest_ancestor_time pair.

Reply:
My question is about the compaction case. Compaction case will have lots of duplication. For example, after a L1->L2, many entries in L2 files are the same, as they come from the same L1 file. And then another L1->L2 comes and these entries need to be dedup. So duplication is common here and my understanding is that right now Sort() is critical to make sure we don't store duplicate data here. Maybe we can even consider to use a std::map?

Reply:
L0 files typically should not have duplicated entries, as assume they do not have seqno overlap.
On the other hand, the data set is pretty small here. The maximum is 100, but typically for higher level files, it's pretty small. For example, if the `preclude_last_level_data_seconds` is set to 3 days, it samples a seqno->time pair every 1 hour, if memtable life-span is less than 1 hour, the map is actually empty for most of L0 files.